File: libraries/python/src/urc_converter/sources/sm/convert.py
"""Mapper from the StepMania source model onto URC charts."""

from itertools import groupby

from ...error import UrcError
from ...model import Chart, Layout, Metadata, Note, NoteType, Version
from .._shared import build_timing, check_hold_overlap, round_ms
from .model import SmChart, SmFile, SmNote, Timing
from .parse import resolve_lanes
from .preprocess import filter_notes, preprocess, rows, warp_intervals

_MEASURE_ROWS = 192
_ROLL_TAP_SPACING_MS = 500

_TYPE_ORDER = {NoteType.N: 0, NoteType.LS: 1, NoteType.LE: 2, NoteType.M: 3, NoteType.F: 4}

_DIFFICULTY_NAMES = {
	"beginner": "Beginner",
	"easy": "Easy",
	"basic": "Easy",
	"light": "Easy",
	"medium": "Medium",
	"another": "Medium",
	"trick": "Medium",
	"standard": "Medium",
	"difficult": "Medium",
	"hard": "Hard",
	"ssr": "Hard",
	"maniac": "Hard",
	"heavy": "Hard",
	"smaniac": "Challenge",
	"challenge": "Challenge",
	"expert": "Challenge",
	"oni": "Challenge",
	"edit": "Edit",
}


def convert_sm(simfile: SmFile) -> list[Chart]:
	"""Convert every chart of a simfile into URC charts."""
	return [_convert_chart(simfile, chart) for chart in simfile.charts]


def _convert_chart(simfile: SmFile, chart: SmChart) -> Chart:
	lanes = resolve_lanes(chart.steps_type)
	timing = chart.timing if chart.timing is not None else simfile.timing
	offset, bpm_segs, stop_segs, warp_segs = preprocess(timing)
	intervals = warp_intervals(warp_segs + timing.warps)
	notes = filter_notes(chart.notes, intervals)

	entries: list[tuple[int, int, str, object]] = []
	for start, dest in intervals:
		entries.append((start, 0, "warp", dest))
	for beat, seconds in timing.delays:
		entries.append((rows(beat), 1, "delay", seconds))
	for index, note in enumerate(notes):
		entries.append((note.row, 2, "note", index))
		if note.tail_row is not None:
			entries.append((note.tail_row, 2, "tail", index))
	for beat, seconds in stop_segs:
		entries.append((rows(beat), 3, "stop", seconds))
	for beat, bpm in bpm_segs:
		entries.append((rows(beat), 4, "bpm", bpm))
	for beat, numerator, denominator in timing.timesigs:
		entries.append((rows(beat), 5, "timesig", (numerator, denominator)))
	for beat, ratio in timing.scrolls:
		entries.append((rows(beat), 6, "scroll", ratio))
	max_row = max((entry[0] for entry in entries), default=0)
	entries.extend(
		(row, 2, "anchor", None) for row in range(0, max_row + _MEASURE_ROWS, _MEASURE_ROWS)
	)

	head_times: list[float] = [0.0] * len(notes)
	tail_times: dict[int, float] = {}
	bpm_points: list[tuple[int, float, int, int]] = []
	sv_points: list[tuple[int, float]] = []
	anchors: list[int] = []

	seconds = -offset
	bpm: float | None = None
	meter = (4, 4)
	multiplier = 1.0
	warping = False
	warp_dest = 0
	prev_row: int | None = None

	entries.sort(key=lambda entry: (entry[0], entry[1]))
	for row, group in groupby(entries, key=lambda entry: entry[0]):
		if prev_row is not None and not warping and bpm is not None:
			seconds += (row - prev_row) / 48.0 * 60.0 / bpm
		prev_row = row
		if warping and row >= warp_dest:
			warping = False

		new_bpm, new_meter, new_multiplier = bpm, meter, multiplier
		for _, _priority, kind, value in group:
			if kind == "warp":
				if warping:
					warp_dest = max(warp_dest, value)
				else:
					warping = True
					warp_dest = value
			elif kind == "delay":
				seconds += value
			elif kind == "note":
				head_times[value] = seconds
			elif kind == "tail":
				tail_times[value] = seconds
			elif kind == "anchor":
				anchors.append(round_ms(seconds * 1000.0))
			elif kind == "stop":
				seconds += value
			elif kind == "bpm":
				new_bpm = value
			elif kind == "timesig":
				new_meter = value
			else:
				new_multiplier = value

		if new_bpm is not None and (new_bpm, new_meter) != (bpm, meter):
			bpm_points.append((round_ms(seconds * 1000.0), new_bpm, *new_meter))
		if new_multiplier != multiplier:
			sv_points.append((round_ms(seconds * 1000.0), new_multiplier))
		bpm, meter, multiplier = new_bpm, new_meter, new_multiplier

	urc_notes = _build_urc_notes(timing, notes, head_times, tail_times)
	first_note_time = min(
		(time for time, _lane, note_type in urc_notes if note_type is not NoteType.LE),
		default=0,
	)

	timing_points = build_timing(
		bpm_points=bpm_points,
		sv_points=sv_points,
		first_note_time=first_note_time,
		source=".sm",
		measure_anchor_ms=next((time for time in anchors if time >= first_note_time), None),
	)

	urc_notes.sort(key=lambda item: (item[0], item[1], _TYPE_ORDER[item[2]]))
	final_notes = [
		Note(timestamp_ms=time - first_note_time, lane=lane, type=note_type)
		for time, lane, note_type in urc_notes
	]
	check_hold_overlap(final_notes)

	title = " ".join(part for part in (simfile.title, simfile.subtitle) if part) or "Unknown"
	return Chart(
		format_version=Version(major=1, minor=1),
		metadata=Metadata(
			original="StepMania",
			title=title,
			artist=simfile.artist or "Unknown",
			creator=chart.credit or simfile.credit or "Unknown",
			version=chart.chartname or _difficulty_name(chart.difficulty, chart.description),
		),
		judgment=None,
		layout=Layout(keys=lanes, special_keys=0, special_lanes=None),
		timing=timing_points,
		notes=final_notes,
	)


def _build_urc_notes(
	timing: Timing,
	notes: list[SmNote],
	head_times: list[float],
	tail_times: dict[int, float],
) -> list[tuple[int, int, NoteType]]:
	fake_ranges = [(rows(beat), rows(beat) + rows(length)) for beat, length in timing.fakes]
	urc_notes: list[tuple[int, int, NoteType]] = []

	for index, note in enumerate(notes):
		head_ms = round_ms(head_times[index] * 1000.0)
		if any(start <= note.row < end for start, end in fake_ranges):
			urc_notes.append((head_ms, note.track, NoteType.F))
		elif note.kind == "hold":
			tail_ms = round_ms(tail_times[index] * 1000.0)
			if tail_ms <= head_ms:
				raise UrcError("syntax", 1, f"hold on lane {note.track} collapses to zero length")
			urc_notes.append((head_ms, note.track, NoteType.LS))
			urc_notes.append((tail_ms, note.track, NoteType.LE))
		elif note.kind == "roll":
			end_ms = round_ms(tail_times[index] * 1000.0)
			urc_notes.append((head_ms, note.track, NoteType.N))
			tap_ms = head_ms + _ROLL_TAP_SPACING_MS
			while tap_ms < end_ms:
				urc_notes.append((tap_ms, note.track, NoteType.N))
				tap_ms += _ROLL_TAP_SPACING_MS
		elif note.kind == "mine":
			urc_notes.append((head_ms, note.track, NoteType.M))
		elif note.kind == "fake":
			urc_notes.append((head_ms, note.track, NoteType.F))
		else:  # tap | lift
			urc_notes.append((head_ms, note.track, NoteType.N))
	return urc_notes


def _difficulty_name(difficulty: str, description: str) -> str:
	name = _DIFFICULTY_NAMES.get(difficulty.strip().lower(), "")
	if name == "Hard" and description.strip().lower() in ("smaniac", "challenge"):
		name = "Challenge"
	return name or "Edit"

File: libraries/dotnet/UrcConverter/Sources/Sm/Convert.fs
namespace UrcConverter.Sources.Sm

module Convert =

    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources
    open UrcConverter.Sources.Sm.Model
    open UrcConverter.Sources.Sm.Notes
    open UrcConverter.Sources.Sm.Parse
    open Preprocess

    [<Literal>]
    let private MeasureRows = 192

    let private typeRank =
        function
        | NoteType.N -> 0
        | NoteType.LS -> 1
        | NoteType.LE -> 2
        | NoteType.M -> 3
        | NoteType.F -> 4


    // --- timing walk --------------------------------------------------------

    type private Entry =
        {
            Row: int
            Priority: int
            Kind: EntryKind
        }

    and private EntryKind =
        | WarpEvent of dest: int
        | DelayEvent of seconds: float
        | NoteEvent of index: int
        | TailEvent of index: int
        | AnchorEvent
        | StopEvent of seconds: float
        | BpmEvent of bpm: float
        | TimeSigEvent of meter: int * int
        | ScrollEvent of ratio: float

    type private WalkState =
        {
            Seconds: float
            Bpm: float option
            Meter: int * int
            Multiplier: float
            Warping: bool
            WarpDest: int
            PrevRow: int option
            HeadTimes: Map<int, float>
            TailTimes: Map<int, float>
            BpmPoints: (int * float * int * int) list
            SvPoints: (int * float) list
            Anchors: int list
        }

    // --- chart conversion ---------------------------------------------------

    let private convertChart (simfile: SmFile) (chart: SmChart) : Result<Chart, UrcError> =
        result {
            let! lanes = resolveLanes chart.StepsType
            let timing = match chart.Timing with Some timing -> timing | None -> simfile.Timing
            let! offset, bpmSegs, stopSegs, warpSegs = preprocess timing
            let intervals = warpIntervals (warpSegs @ timing.Warps)
            let notes = filterNotes chart.Notes intervals

            let entries =
                [
                    for start, dest in intervals do
                        yield { Row = start; Priority = 0; Kind = WarpEvent dest }

                    for beat, seconds in timing.Delays do
                        yield { Row = rowsOf beat; Priority = 1; Kind = DelayEvent seconds }

                    for index, note in List.indexed notes do
                        yield { Row = note.Row; Priority = 2; Kind = NoteEvent index }

                        match note.TailRow with
                        | Some tailRow -> yield { Row = tailRow; Priority = 2; Kind = TailEvent index }
                        | None -> ()

                    for beat, seconds in stopSegs do
                        yield { Row = rowsOf beat; Priority = 3; Kind = StopEvent seconds }

                    for beat, bpm in bpmSegs do
                        yield { Row = rowsOf beat; Priority = 4; Kind = BpmEvent bpm }

                    for beat, numerator, denominator in timing.TimeSignatures do
                        yield { Row = rowsOf beat; Priority = 5; Kind = TimeSigEvent(numerator, denominator) }

                    for beat, ratio in timing.Scrolls do
                        yield { Row = rowsOf beat; Priority = 6; Kind = ScrollEvent ratio }
                ]

            let maxRow = entries |> List.map (fun entry -> entry.Row) |> List.fold max 0

            let anchorEntries =
                [
                    for row in 0 .. MeasureRows .. maxRow + MeasureRows - 1 ->
                        { Row = row; Priority = 2; Kind = AnchorEvent }
                ]

            let entries =
                anchorEntries @ entries |> List.sortBy (fun entry -> entry.Row, entry.Priority)

            let initial =
                {
                    Seconds = -offset
                    Bpm = None
                    Meter = (4, 4)
                    Multiplier = 1.0
                    Warping = false
                    WarpDest = 0
                    PrevRow = None
                    HeadTimes = Map.empty
                    TailTimes = Map.empty
                    BpmPoints = []
                    SvPoints = []
                    Anchors = []
                }

            let walk (state: WalkState) (row: int, evts: Entry list) : WalkState =
                let advanced =
                    match state.PrevRow, state.Bpm with
                    | Some prevRow, Some bpm when not state.Warping ->
                        {
                            state with
                                Seconds = state.Seconds + float (row - prevRow) / 48.0 * 60.0 / bpm
                        }
                    | _ -> state

                let entered =
                    if advanced.Warping && row >= advanced.WarpDest then
                        { advanced with Warping = false }
                    else
                        advanced

                let grouped =
                    evts
                    |> List.fold
                        (fun (state: WalkState) (entry: Entry) ->
                            match entry.Kind with
                            | WarpEvent dest ->
                                if state.Warping then
                                    { state with WarpDest = max state.WarpDest dest }
                                else
                                    { state with Warping = true; WarpDest = dest }
                            | DelayEvent seconds -> { state with Seconds = state.Seconds + seconds }
                            | NoteEvent index ->
                                { state with HeadTimes = Map.add index state.Seconds state.HeadTimes }
                            | TailEvent index -> { state with TailTimes = Map.add index state.Seconds state.TailTimes }
                            | AnchorEvent ->
                                { state with Anchors = Shared.roundMs (state.Seconds * 1000.0) :: state.Anchors }
                            | StopEvent seconds -> { state with Seconds = state.Seconds + seconds }
                            | BpmEvent bpm -> { state with Bpm = Some bpm }
                            | TimeSigEvent(numerator, denominator) ->
                                { state with Meter = (numerator, denominator) }
                            | ScrollEvent ratio -> { state with Multiplier = ratio })
                        entered

                let bpmChanged =
                    match grouped.Bpm with
                    | Some bpm when grouped.Bpm <> state.Bpm || grouped.Meter <> state.Meter ->
                        let numerator, denominator = grouped.Meter
                        (Shared.roundMs (grouped.Seconds * 1000.0), bpm, numerator, denominator) :: grouped.BpmPoints
                        |> Some
                    | _ -> None

                let svChanged =
                    if grouped.Multiplier <> state.Multiplier then
                        Some((Shared.roundMs (grouped.Seconds * 1000.0), grouped.Multiplier) :: grouped.SvPoints)
                    else
                        None

                {
                    grouped with
                        BpmPoints = bpmChanged |> Option.defaultValue grouped.BpmPoints
                        SvPoints = svChanged |> Option.defaultValue grouped.SvPoints
                        PrevRow = Some row
                }

            let walked =
                entries
                |> List.groupBy (fun entry -> entry.Row)
                |> List.fold walk initial

            let! urcNotes = buildUrcNotes timing notes walked.HeadTimes walked.TailTimes

            let firstNoteTime =
                match
                    urcNotes
                    |> List.filter (fun (_, _, kind) -> kind <> NoteType.LE)
                    |> List.map (fun (time, _, _) -> time)
                with
                | [] -> 0
                | times -> List.min times

            let anchors = List.rev walked.Anchors

            let! timingPoints =
                Shared.buildTiming
                    (List.rev walked.BpmPoints)
                    (List.rev walked.SvPoints)
                    firstNoteTime
                    ".sm"
                    (anchors |> List.tryFind (fun time -> time >= firstNoteTime))

            let ordered =
                urcNotes
                |> List.sortBy (fun (time, lane, kind) -> time, lane, typeRank kind)
                |> List.map (fun (time, lane, kind) ->
                    {
                        TimestampMs = time - firstNoteTime
                        Lane = lane
                        Type = kind
                    })

            do! Shared.checkHoldOverlap ordered

            let title =
                [ simfile.Title; simfile.Subtitle ]
                |> List.filter (fun part -> part <> "")
                |> String.concat " "

            let creator = firstNonEmpty chart.Credit simfile.Credit "Unknown"

            let version =
                if chart.ChartName <> "" then
                    chart.ChartName
                else
                    difficultyName chart.Difficulty chart.Description

            return
                {
                    FormatVersion = { Major = 1; Minor = 1 }
                    Metadata =
                        {
                            Original = "StepMania"
                            Title = if title = "" then "Unknown" else title
                            Artist = if simfile.Artist = "" then "Unknown" else simfile.Artist
                            Creator = creator
                            Version = version
                        }
                    Judgment = None
                    Layout = { Keys = lanes; SpecialKeys = 0; SpecialLanes = None }
                    TimingPoints = timingPoints
                    Notes = ordered
                }
        }

    /// Converts every chart of a simfile into URC charts.
    let convertSm (simfile: SmFile) : Result<Chart list, UrcError> =
        simfile.Charts |> List.traverseResultM (convertChart simfile)
